MITLibraries/timdex-pipeline-lambdas
# Setup initial auto-deploy for Lambda Functions

See the commit for a detailed explanation of what was added and why. The rest of this PR is overview of what should happen and how to do some testing/verification.

## Overview of auto-deploy process

The flow of actions is `local gha caller workflow` --> `shared workflow` --> `make commands in local Makefile`.

That is, when the `dev-build.yml` workflow is triggered by a PR into `main`, 
1. it calls the `lambda-shared-deploy.yml` workflow and passes to it the name of the oidc role, the AWS region, and the account number of Dev1 account. 
2. The `lambda-shared-deploy.yml` workflow handles the oidc auth to AWS, sets some basic environment variables for the GitHub runner, and then executes the `make` commands.
3. The `Makefile` sets some additional environment variables when it runs (pulling values from Param Store) so that it is accessing the correct AWS account with the correct role (and finding the correct ECR repository and the correct Lambda function)

## Workflow update

There is a new workflow [dev-build.yml](.github/workflows/dev-build.yml) that is currently configured with two triggers: 
* `workflow_dispatch`: this is here so that a review can test the functionality of the workflow by running it manually in the Actions tab on this repo
* `on.pull_request.branches: main`: When this PR is created, this workflow should (should have?) run and pushed a new version of the container to the ECR and updated the Lambda function to use the new version of the container.

## Makefile update

There is a greatly expanded Makefile. The existing commands are still there (but slightly modified to force them to only work in the Dev1 account). Additional commands were put in that should only be used by GitHub Actions. These are called by the shared workflow found in the [container-flows-ssm](https://github.com/MITLibraries/.github/tree/container-flows-ssm) branch in our shared `.github` repository.

## How to Verify/Review

To verify that the automation worked, there are two places to look:
1. In the Actions tab of the repo, you should find the triggered run. You can/should dig into the details of that run to verify that all the steps worked and that there was nothing funny going on or any secret information revealed.
4. In the Dev1 AWS Account, you should go in to the AWS Console and review both the Lambda function itself (to see what version of the container it is running) and the ECR repository to see the updated container.

The link below should take you straight to the ECR Repository (`timdex-pipeline-lambdas-dev`) to see the list of the last five container versions (**NOTE: You need to log in to AWS SSO and then log in to the Dev1 account before this link will work**):
https://us-east-1.console.aws.amazon.com/ecr/repositories/private/222053980223/timdex-pipeline-lambdas-dev?region=us-east-1

The link below should take you straight to the Lambda function (`timdex-format-dev`) to see what version of the container is currently loaded (**NOTE: You need to log in to AWS SSO and then log in to the Dev1 account before this link will work**):
https://us-east-1.console.aws.amazon.com/lambda/home?region=us-east-1#/functions/timdex-format-dev?tab=code

You should also verify that you can still `docker push` and `lambda update` from your local computer by executing your typical `make dist-dev`, `make publish-dev`, and `make update-lambda-dev` commands (after logging in to AWS in your terminal session).

## Final comment 

This is just a first step -- there is no automation configured yet for a merge to `main` nor for a tagged release on `main`. This is because we have **not** built the necessary infrastructure in Stage-Workloads or Prod-Workloads to support this yet.
